src/libos/crates/async-sfs/src/fs.rs

```diff
@@ -0,0 +1,1367 @@
+            }
+        }
+        Ok(())
+    }
+
+    /// Read content, no matter what type it is
+    async fn _read_at(&self, offset: usize, buf: &mut [u8]) -> Result<usize> {
+        let file_size = self.disk_inode.size as usize;
+        let iter = BlockRangeIter {
+            begin: file_size.min(offset),
+            end: file_size.min(offset + buf.len()),
+            block_size: BLOCK_SIZE,
+        };
+
+        let mut read_len = 0;
+        for range in iter {
+            let device_block_id = self.get_device_block_id(range.block_id).await?;
+            let len = self
+                .fs()
+                .storage
+                .read_at(
+                    device_block_id,
+                    &mut buf[read_len..read_len + range.len()],
+                    range.begin,
+                )
+                .await?;
+            debug_assert!(len == range.len());
+            read_len += len;
+        }
+
+        Ok(read_len)
+    }
+
+    /// Write content, no matter what type it is
+    async fn _write_at(&self, offset: usize, buf: &[u8]) -> Result<usize> {
+        let file_size = self.disk_inode.size as usize;
+        let iter = BlockRangeIter {
+            begin: file_size.min(offset),
+            end: file_size.min(offset + buf.len()),
+            block_size: BLOCK_SIZE,
+        };
+
+        let mut write_len = 0;
+        for range in iter {
+            let device_block_id = self.get_device_block_id(range.block_id).await?;
+            let len = self
+                .fs()
+                .storage
+                .write_at(
+                    device_block_id,
+                    &buf[write_len..write_len + range.len()],
+                    range.begin,
+                )
+                .await?;
+            debug_assert!(len == range.len());
+            write_len += len;
+        }
+
+        Ok(write_len)
+    }
+
+    fn nlinks_inc(&mut self) {
+        self.disk_inode.nlinks += 1;
+    }
+
+    fn nlinks_dec(&mut self) {
+        assert!(self.disk_inode.nlinks > 0);
+        self.disk_inode.nlinks -= 1;
+    }
+
+    fn dirty(&self) -> bool {
+        self.disk_inode.dirty()
+    }
+
+    async fn sync_data(&self) -> Result<()> {
+        if let Some(cache) = self.fs().storage.as_page_cache() {
+            let data_pages = {
+                let mut data_pages = Vec::new();
+                for id in 0..self.disk_inode.blocks as BlockId {
+                    let device_block_id = self.get_device_block_id(id).await?;
+                    data_pages.push(device_block_id);
+                }
+                data_pages
+            };
+            cache.flush_pages(&data_pages).await?;
+        }
+        Ok(())
+    }
+
+    async fn sync_metadata(&mut self) -> Result<()> {
+        // The inode is freed already, do nothing
+        if self.is_freed {
+            return Ok(());
+        }
+
+        // nlinks is zero, just delete it
+        if self.disk_inode.nlinks == 0 {
```

Comment:
If the `nlinks == 0`, he inode is deleted. When we close a file, it will call the `sync_metadata` method, and we will have a chance to delete that inode in fs.